libcloud/compute/drivers/gridscale.py

```diff
@@ -0,0 +1,977 @@
+
+        ip = self.ex_create_ip(4, location, name + '_ip')
+
+        self.attach_volume(node, volume)
+        self.ex_link_ip_to_node(node, ip)
+        self.ex_link_network_to_node(node, self.ex_list_networks()[0])
+        self.ex_start_node(node)
+
+        return self._to_node(self._get_resource('servers', node.id))
+
+    def ex_create_ip(self, family, location, name):
+        """
+        Create either an ip_v4 ip or a ip_v6.
+
+        :param family: Defines if the ip is v4 or v6 with int 4 or int 6.
+        :type family: ``int``
+
+        :param location: Defines which datacenter the created ip
+                         responds with.
+        :type location: :class:`.NodeLocation`
+
+        :param name: Name of your Ip.
+        :type name: ``str``
+
+        :return: Ip
+        :rtype: :class:`.GridscaleIp`
+        """
+        self.connection.async_request(
+            'objects/ips/',
+            data=json.dumps({
+                'name': name,
+                'family': family,
+                'location_uuid': location.id}),
+            method='POST')
+
+        return self._to_ip(self._get_resource('ips', self.connection
+                                              .poll_response_initial
+                                              .object['object_uuid']))
+
+    def ex_create_networks(self, name, location):
+        """
+        Create a network at the data center location.
+
+        :param name: Name of the network.
+        :type name: ``str``
+
+        :param location: Location.
+        :type location: :class:`.NodeLocation`
+
+        :return: Network.
+        :rtype: :class:`.GridscaleNetwork`
+        """
+        self.connection.async_request(
+            'objects/networks',
+            data=json.dumps({
+                'name': name,
+                'location_uuid': location.id}),
+            method='POST')
+
+        return self._to_network(self._get_resource('network', self.connection
+                                                   .poll_response_initial
+                                                   .object['object_uuid']))
+
+    def create_volume(self, size, name, location=None, snapshot=None):
+        """
+        Create a new volume.
+
+        :param size: Integer in GB.
+        :type size: ``int``
+
+        :param name: Name of the volume.
+        :type name: ``str``
+
+        :param location: The server location.
+        :type location: :class:`.NodeLocation`
+
+        :param snapshot:  Snapshot from which to create the new
+                          volume.  (optional)
+        :type snapshot: :class:`.VolumeSnapshot`
+
+        :return: Newly created StorageVolume.
+        :rtype: :class:`.StorageVolume`
+        """
+
+        return self._create_volume_from_template(size, name, location)
+
+    def _create_volume_from_template(self, size, name, location=None,
+                                     template={}):
```

Comment:
Mutable defaults are not really safe in Python - https://docs.python-guide.org/writing/gotchas/#mutable-default-arguments

Please default it to ``None`` and do something along the lines of ``template = template or {}`` inside the method body.